[inductor][overlap] Config to log final collectives runtime estimations to compare with profile

IvanKobzarev opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #170120

We already have logging before overlap/bucketing.
But this we can not compare with profile that will have collectives after overlap and bucketing.

Adding this ability to verify logged collectives in the order how we they see them in profile.



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

IvanKobzarev committed:
[inductor][overlap] Config to log final collectives runtime estimations to compare with profile

[ghstack-poisoned]

eellison commented on torch/_inductor/fx_passes/node_runtime_estimation.py:
```diff
@@ -256,20 +257,86 @@ def _node_summary(n: fx.Node) -> str:
+            success, args, kw = fx_utils.get_fake_args_kwargs(start)
+            if success:
+
+                def extract_first_tensor_bytes(t: torch.Tensor) -> torch.Tensor:
```
does this function already exist somewhere else ?

eellison commented on torch/_inductor/fx_passes/overlap_scheduling.py:
```diff
@@ -289,6 +289,7 @@ def __init__(
         collective_estimator: Literal["analytical", "benchmark"],
         max_memory_increase_gb: float | None = 1.0,
         max_memory_increase_ratio: float | None = 0.05,
+        log_final_collectives_estimations: bool = False,
```
since this doesnt affect functionality, just logging, maybe we remove it as a top level arg and just config it somewhere.

IvanKobzarev replied:
Yeah, I worried about the cases when OverlapScheduler will be forked and global deps will need some afterwork and decided to avoid them.

eellison reviewed (approved):
Looks good, just one comment about not putting as a top level arg, i dont think we want to have to maintain bc etc for debugging configs

IvanKobzarev commented on torch/_inductor/fx_passes/node_runtime_estimation.py:
```diff
@@ -303,7 +370,7 @@ def _log_collective_benchmarks(
         rows.append(
             [
-                key[:80],
+                key[:100],
```
It's a title with size bytes, group name etc. I will log it all.

IvanKobzarev committed:
Update on "[inductor][overlap] Config to log final collectives runtime estimations to compare with profile"


We already have logging before overlap/bucketing.
But this we can not compare with profile that will have collectives after overlap and bucketing.

Adding this ability to verify logged collectives in the order how we they see them in profile.



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo

[ghstack-poisoned]
